File: problems/tree/0100-same-tree.js
/**
 * #100 Same Tree
 * Pattern: compare matching positions in both trees recursively.
 * Time: O(n) | Space: O(h) call stack
 *
 * @param {TreeNode|null} p
 * @param {TreeNode|null} q
 * @return {boolean}
 */
function isSameTree(p, q) {
  if (p === null || q === null) return p === q
  if (p.val !== q.val) return false
  return isSameTree(p.left, q.left) && isSameTree(p.right, q.right)
}

module.exports = isSameTree

File: problems/tree/0100-same-tree.py
"""#100 Same Tree
Compare node values and corresponding left and right subtrees recursively.
Time: O(n) | Space: O(h) call stack.
"""


class Solution:
    def isSameTree(self, p, q):
        if p is None or q is None:
            return p is q
        if p.val != q.val:
            return False
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
